### PR TITLE
[BEAM-618] Disallow NAN, INF and -INF invalid JSON values in bigquery exporter

### PR DESCRIPTION
**Pull request by ajamato**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

Now exporting JSON will fail with invalid NAN, INF or -INF values.

### CODE REVIEW

**Comment by ajamato** [comment]
R: @chamikaramj

File: sdks/python/apache_beam/io/bigquery_test.py [hunk 1]
```diff
@@ -115,6 +176,33 @@ def test_simple_schema_as_json(self):
             {'name': 'n', 'type': 'INTEGER', 'mode': 'NULLABLE'}]}),
         sink.schema_as_json())
 
+  def test_schema_as_json_invalid_nan(self):
+    sink = beam.io.BigQuerySink(
+        'dataset.table', schema='s:STRING, n:INTEGER')
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/bigquery_test.py]: How are these three tests different from 'test_simple_schema_as_json' ?

**Comment by chamikaramj** [comment]
Seems like there are lint errors.

File: sdks/python/apache_beam/io/bigquery.py [hunk 2]
```diff
@@ -145,7 +146,14 @@ class RowAsDictJsonCoder(coders.Coder):
   """
 
   def encode(self, table_row):
-    return json.dumps(table_row)
+    # The normal error when dumping NAN/INF values is:
+    # ValueError: Out of range float values are not JSON compliant
+    # This code will catch this error to emit an error that explains
+    # to the programmer that they have used NAN/INF values.
+    try:
+      return json.dumps(table_row, allow_nan=False)
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/bigquery.py]: Do you know the performance impact of this change ?

File: sdks/python/apache_beam/io/bigquery_test.py [hunk 3]
```diff
@@ -40,6 +40,27 @@ def test_row_as_dict(self):
     test_value = {'s': 'abc', 'i': 123, 'f': 123.456, 'b': True}
     self.assertEqual(test_value, coder.decode(coder.encode(test_value)))
 
+  def test_invalid_json_nan(self):
+    with self.assertRaises(ValueError) as exn:
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/bigquery_test.py]: Also make sure that we are getting the correct error message.

File: sdks/python/apache_beam/io/bigquery.py (hunk 2 above)
Comment (ajamato) [inline sdks/python/apache_beam/io/bigquery.py]: I imagine whatever code they use to detect NAN and serialize it out will have to raise an exception instead. So I don't expect there to be anything noticeable.  Is there a benchmark or something you'd like me to run?

File: sdks/python/apache_beam/io/bigquery_test.py (hunk 3 above)
Comment (ajamato) [inline sdks/python/apache_beam/io/bigquery_test.py]: That's done here self.assertTrue(bigquery.JSON_COMPLIANCE_ERROR in exn.exception.message). Do you want to make sure that the base error is what we expect? We only catch the ValueError, I could also inspect the string of the ValueError in bigquery.py to make sure I only raise when it matches that.

File: sdks/python/apache_beam/io/bigquery_test.py (hunk 1 above)
Comment (ajamato) [inline sdks/python/apache_beam/io/bigquery_test.py]: meant to rm these

**Comment by ajamato** [comment]
Bare with me, as I learn this. Making some changes still

File: sdks/python/apache_beam/io/bigquery.py (hunk 2 above)
Comment (chamikaramj) [inline sdks/python/apache_beam/io/bigquery.py]: We could try running bigquery_tornadoes example using DataflowPipelineRunner. If you don't see a significant performance difference there we should be good to go. I was just worried if this adds some significantly slow per-element check.

File: sdks/python/apache_beam/io/bigquery_test.py (hunk 3 above)
Comment (chamikaramj) [inline sdks/python/apache_beam/io/bigquery_test.py]: This looks good.

**Comment by chamikaramj** [comment]
LGTM

**Comment by ajamato** [comment]
R: @dhalperi Hi Dan, would you please merge in this change. Thanks.

File: sdks/python/apache_beam/io/bigquery_test.py (hunk 1 above)
Comment (dhalperi) [inline sdks/python/apache_beam/io/bigquery_test.py]: Looks like they're still here. A bug?

Comment (ajamato) [inline sdks/python/apache_beam/io/bigquery_test.py]: My bad, rm'd these tests now

File: sdks/python/apache_beam/io/bigquery_test.py (hunk 3 above)
Comment (chamikaramj) [inline sdks/python/apache_beam/io/bigquery_test.py]: Refactor these tests to reduce number of lines ?

Comment (ajamato) [inline sdks/python/apache_beam/io/bigquery_test.py]: Done

**Comment by dhalperi** [comment]
LGTM, merging.

**Comment by dhalperi** [comment]
Merged; please close at your leisure.

**Pull request closed without merging**